adapters/mysql2/lib/opentelemetry/adapters/mysql2/patches/client.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Adapters
+    module Mysql2
+      module Patches
+        # Module to prepend to Mysql2::Client for instrumentation
+        module Client
+          def query(sql, options = {})
+            response = nil
+
+            tracer.in_span(
+              database_span_name,
+              attributes: client_attributes.merge(
+                'db.statement' => sql
+              ),
+              kind: :client
+            ) do
+              response = super(sql, options)
+            end
+
+            response
+          end
+
+          private
+
+          def database_span_name
+            # Without obfuscation or quantization of SQL Queries, setting span name
+            # To the sql query woud result in PII or Cardinality issues
+            # Current Otel apprroach seems to be {database.component_name}.{database_instance_name}
+            # https://github.com/open-telemetry/opentelemetry-python/blob/39fa078312e6f41c403aa8cad1868264011f7546/ext/opentelemetry-ext-dbapi/tests/test_dbapi_integration.py#L53
+            # This would create span names like mysql.default, mysql.replica, postgresql.staging etc etc
+            database_name ? "mysql.#{database_name}" : 'mysql'
+          end
```

Comment:
The [semantic conventions for databases](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md) states:

> Span `name` should be set to low cardinality value representing the statement executed on the database. It may be stored procedure name (without argument), sql statement without variable arguments, etc. When it's impossible to get any meaningful representation of the span `name`, it can be populated using the same value as `db.instance`.

Do we have any way to get a meaningful, low-cardinality representation of the statement at this point? E.g. “stored procedure name (without argument), sql statement without variable arguments”. I *think* the answer is “no”, in which case what you have seems fine, but I’d love to be proven wrong.

Reply:
yea, so i poked around the mysql2 codebase a good amount here, and also looked into how some similar open-telemetry-python instrumentations were doing things, and it seems like the answer is no. That being said I am very far from a SQL expert, so i'd be curious to know how others are handling this. Is it just that everyone is breaking spec?  Additionally, how are other's handling the `db.statement` sql obfuscation, is it all at the collector level? If there's an example of obfuscation of db.statement done at the tracer adapter level, perhaps we could leverage that approach.

Reply:
> how are other's handling the `db.statement` sql obfuscation

I don’t think anyone is. We (Shopify) dug into this quite a bit a few years ago and concluded we’d have to add a SQL parser to sanitize or obfuscate the query. **We** redact the `db.statement` attribute from all spans in the collector — there is no obfuscation support specific to SQL queries there — but I don’t know what others are doing.

Reply:
I see, that makes sense. From the perspective of where this instrumentation is being ported from (dd-trace-rb), the `datadog-agent` [handles the obfuscation](https://github.com/DataDog/datadog-agent/blob/master/pkg/trace/obfuscate/sql.go), which makes low cardinality representations of sql possible without adding that work into the tracing client code. In this case, I don't believe it's possible to set span.name to a representation of sql.

Reply:
I don't think we can use the raw sql as a span name, but I join you all in not knowing the best way to deal with this. The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md) says:

> When it's impossible to get any meaningful representation of the span name, it can be populated using the same value as db.instance.

However, that doesn't make this instrumentation very useful it all it does is create a span with the database instance information. I'd like to see if I can find any other SIGs who have encountered problem, but haven't yet. I'll look / ask around a bit more.

One possible middle ground I can think of, is that we could try to identify the statement type and use it for the span name. That would give us span names such as "MySQL SELECT", "MySQL INSERT", etc. This would be similar to http client instrumentation using the verb when it can't derive a low cardinality name. We should discuss if this idea has any merit before moving forward with it.

Reply:
> One possible middle ground I can think of, is that we could try to identify the statement type and use it for the span name. That would give us span names such as "MySQL SELECT", "MySQL INSERT", etc. This would be similar to http client instrumentation using the verb when it can't derive a low cardinality name. We should discuss if this idea has any merit before moving forward with it.

Coincidentally, we just made a very similar change to our Shopify-internal tracing gem. It seems like a reasonable path forward.

Reply:
sure, i agree that `MySQL <INSERT/SELECT/DELETE/etc>` seems reasonable enough. I can think of some very inefficient ways to do this but going look to see if this pattern is being used in other otel repo's so that I don't have to re-invent the wheel here

Reply:
Our code for this identifies specific patterns:
```ruby
        QUERY_NAMES = [
          "set names",
          "select",
          "insert",
          "update",
          "delete",
          "begin",
          "commit",
          "rollback",
          "savepoint",
          "release savepoint",
          "explain",
          "drop database",
          "drop table",
          "create database",
          "create table",
        ].freeze

        QUERY_NAME_RE = Regexp.new("^(#{QUERY_NAMES.join('|')})", Regexp::IGNORECASE)

...
          QUERY_NAME_RE.match(sql) { |match| match[1].downcase } unless sql.nil?
```

Reply:
that's extremely helpful, thank you. I've updated the PR and tests accordingly